minio/select/reader.py

```diff
@@ -0,0 +1,229 @@
+    """
+    def __init__(self, response):
+        self.response = response
+        self.remaining_bytes = bytes()
+        self.stat = {}
+        self.prog = {}
+
+    def readable(self):
+        return True
+
+    def writeable(self):
+        return False
+
+    def close(self):
+        self.response.close()
+
+    def stats(self):
+        return self.stat
+
+    def progress(self):
+        return self.prog
+
+    def __extract_message(self):
+        """
+        Process the response sent from server.
+        https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectSELECTContent.html
+        """
+
+        crc_bytes = io.BytesIO()
+        total_bytes_len = self.response.read(4)
+        if len(total_bytes_len) == 0:
+            return {}
+
+        total_length = byte_int(total_bytes_len)
+        header_bytes_len = self.response.read(4)
+        if len(header_bytes_len) == 0:
+            return {}
+
+        header_len = byte_int(header_bytes_len)
+
+        crc_bytes.write(total_bytes_len)
+        crc_bytes.write(header_bytes_len)
+
+        prelude_bytes_crc = self.response.read(4)
+        if not validate_crc(crc_bytes.getvalue(), prelude_bytes_crc):
+            raise SelectCRCValidationError(
+                {"Checksum Mismatch, PreludeCRC of " +
+                 str(calculate_crc(crc_bytes.getvalue())) +
+                 " does not equal expected CRC of " +
+                 str(byte_int(prelude_bytes_crc))})
+
+        crc_bytes.write(prelude_bytes_crc)
+
+        header_bytes = self.response.read(header_len)
+        if len(header_bytes) == 0:
+            raise SelectMessageError(
+                "Premature truncation of select message header"+
+                ", server is sending corrupt message?")
+
+        crc_bytes.write(header_bytes)
+
+        header_map = _extract_header(header_bytes)
+        payload_length = total_length - header_len - int(16)
+        payload_bytes = b''
+        event_type = header_map["event-type"]
+        if header_map["message-type"] == ERROR:
+            raise SelectMessageError(
+                header_map["error-code"] + ":\"" + \
+                header_map["error-message"] + "\"")
+        elif header_map["message-type"] == EVENT:
+            if event_type == EVENT_END:
+                pass
+            elif event_type == EVENT_CONT:
+                pass
+            elif event_type == EVENT_STATS:
+                content_type = header_map["content-type"]
+                if content_type != EVENT_CONTENT_TYPE:
+                    raise SelectMessageError(
+                        "Unrecognized content-type {0}".format(content_type))
+                else:
+                    payload_bytes = self.response.read(payload_length)
+                    self.stat = _parse_stats(payload_bytes)
+
+            elif event_type == EVENT_RECORDS:
+                payload_bytes = self.response.read(payload_length)
+        else:
+            raise SelectMessageError(
```

Comment:
We know the server response ,so do you think we can ever get this `unrecognised message type`

Reply:
Yes because server can prematurely, close the connection. Short reads are common.